[Supplier] Initial slashing implementation

red-0ne opened the pull request:
## Summary

This PR introduces the initial implementation of `Supplier` slashing:
- Adjusts proof parameters to align with the probabilistic proofs paper.
- Adds comments on potential follow-up improvements and adjustments.
- Implements `Supplier` slashing for invalid or missing proofs.
- Unstakes a `Supplier` whose stake falls below the minimum due to slashing.
- Move the `ComputeUnitsToTokenMultiplier` to the `shared` module.

_~1300LOC is auto-generated code_

## Issue

To enable permissionless demand, the protocol must discourage the submission of false or invalid claims.

As part of this, `Suppliers` should be penalized for failing to comply with the C&P life cycle rules.

- #758 

## Type of change

Select one or more from the following:

- [x] New feature, functionality or library
- [x] Consensus breaking; add the `consensus-breaking` label if so. See #791 for details
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [ ] **Documentation**: `make docusaurus_start`; only needed if you make doc changes
- [x] **Unit Tests**: `make go_develop_and_test`
- [x] **LocalNet E2E Tests**: `make test_e2e`
- [x] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have commented my code
- [x] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [x] I have left TODOs throughout the codebase, if applicable

Olshansk reviewed (changes requested):
Haven't gone through the meet yet in `x/tokenomics/keeper/settle_pending_claims.go` but publishing a partial review

red-0ne commented on x/proof/types/params.go:
```diff
@@ -29,13 +29,19 @@ var (
 	ParamProofRequestProbability           = "proof_request_probability"
 	DefaultProofRequestProbability float32 = 0.25 // See: https://github.com/pokt-network/pocket-core/blob/staging/docs/proposals/probabilistic_proofs.md
 
+	// TODO_BETA(@red-0ne): Change the proofRequirementThreshold to be of type uPOKT coin.
+	// The probabilistic proofs paper specifies a threshold of 20 POKT.
+	// We should change this param's name to ProofRequirementThresholduPOKT and type
+	// to uPOKT coin with a default value of 20e6 uPOKT.
```
Added in this PR.

red-0ne commented on pkg/relayer/session/claim.go:
```diff
@@ -181,6 +181,11 @@ func (rs *relayerSessionsManager) newMapClaimSessionsFn(
 		}
 
 		// Map key is the supplier operator address.
+		// TODO_BETA(@red-0ne, #780, #765): Ensure that the supplier operator account
```
This is trivial to add, will do it in this PR.

red-0ne commented on x/tokenomics/keeper/token_logic_modules.go:
```diff
@@ -564,7 +564,13 @@ func (k Keeper) ensureClaimAmountLimits(
 }
 
-// numRelaysToCoin calculates the amount of uPOKT to mint based on the number of compute units.
+// TODO_BETA(@red-0ne, #745): Once #745 is merged rename numRelaysToCoin to
```
Implemented

Olshansk commented on x/tokenomics/keeper/settle_pending_claims.go:
```diff
@@ -305,15 +331,97 @@ func (k Keeper) proofRequirementForClaim(ctx sdk.Context, claim *prooftypes.Clai
+	logger := k.logger.With("method", "slashSupplierStake")
+
+	proofParams := k.proofKeeper.GetParams(ctx)
+	slashingPenalty := proofParams.GetProofMissingPenalty()
```
```suggestion
	slashingPenaltyPerExpiredClaim := proofParams.GetProofMissingPenalty()
```

Olshansk commented on x/tokenomics/keeper/settle_pending_claims.go:
```diff
@@ -305,15 +331,97 @@ func (k Keeper) proofRequirementForClaim(ctx sdk.Context, claim *prooftypes.Clai
+	remainingStakeAmt := math.NewInt(0)
+	if supplierToSlash.GetStake().Amount.GT(totalSlashingAmt) {
+		remainingStakeAmt = supplierToSlash.GetStake().Amount.Sub(totalSlashingAmt)
+	}
```
Please log a warning w/ details in the else case

Olshansk commented on x/tokenomics/keeper/settle_pending_claims.go:
```diff
@@ -305,15 +331,97 @@ func (k Keeper) proofRequirementForClaim(ctx sdk.Context, claim *prooftypes.Clai
+		remainingStakeAmt = supplierToSlash.GetStake().Amount.Sub(totalSlashingAmt)
+	}
+
+	err := k.bankKeeper.MintCoins(ctx, types.ModuleName, sdk.NewCoins(totalSlashingCoin))
```
s/types/tokenomictypes to make sure there's no ambiguity

Olshansk replied:
Instead of minting, what we need to do is move tokens from the supplier module to the tokenomics module. Otherwise accounting doesn't work.

#PUC at the top as well

Olshansk commented on x/tokenomics/keeper/settle_pending_claims.go:
```diff
@@ -305,15 +331,97 @@ func (k Keeper) proofRequirementForClaim(ctx sdk.Context, claim *prooftypes.Clai
+	}
+
+	err := k.bankKeeper.MintCoins(ctx, types.ModuleName, sdk.NewCoins(totalSlashingCoin))
+	if err != nil {
```
single line w/ the above

Olshansk commented on x/tokenomics/keeper/settle_pending_claims.go:
```diff
@@ -305,15 +331,97 @@ func (k Keeper) proofRequirementForClaim(ctx sdk.Context, claim *prooftypes.Clai
+		return err
+	}
+
+	supplierToSlash.Stake.Amount = remainingStakeAmt
```
Please just create a new stake object

Olshansk commented on x/tokenomics/keeper/settle_pending_claims.go:
```diff
@@ -305,15 +331,97 @@ func (k Keeper) proofRequirementForClaim(ctx sdk.Context, claim *prooftypes.Clai
+
+	logger.Warn(fmt.Sprintf(
+		"slashing supplier %s stake by %s, remaining stake: %s",
+		supplierToSlash.GetOperatorAddress(),
```
Are you sure it's the operator and not the owner we're slashing here?

red-0ne replied:
Will update to include both, It's true that the owner is the one being slashed, but knowing which operator led to slashing is important info too.

Olshansk commented on x/tokenomics/keeper/settle_pending_claims.go:
```diff
@@ -305,15 +331,97 @@ func (k Keeper) proofRequirementForClaim(ctx sdk.Context, claim *prooftypes.Clai
+	// at session end height so the unstaked supplier may not be immediately removed.
+	// Ensure that Gateways and Applications do not interact with a supplier that
+	// is below the minimum stake.
+	if remainingStakeAmt.LT(math.NewInt(1)) {
```
move `math.NewInt(1)` to a local variable w/ a TODO there.

Let's assign it to @bryanchriswhite actually.

Olshansk commented on x/tokenomics/keeper/settle_pending_claims.go:
```diff
@@ -305,15 +331,97 @@ func (k Keeper) proofRequirementForClaim(ctx sdk.Context, claim *prooftypes.Clai
+
+	// Check if the supplier's stake is below the minimum and unstake it if necessary.
+	// TODO_BETA(@red-0ne, #612): Use minimum stake governance parameter once available.
+	// TODO_BETA(@red-0ne): Since SettlePendingClaims is not necessarily called
```
I think it's fine to have this be a TODO_MAINNET

Olshansk commented on x/tokenomics/keeper/settle_pending_claims.go:
```diff
@@ -305,15 +331,97 @@ func (k Keeper) proofRequirementForClaim(ctx sdk.Context, claim *prooftypes.Clai
+		unstakeSessionEndHeight := uint64(shared.GetSessionEndHeight(&sharedParams, currentHeight))
+
+		logger.Warn(fmt.Sprintf(
+			"unstaking supplier %s due to stake below the minimum",
```
Log the min

Olshansk commented on x/tokenomics/keeper/settle_pending_claims.go:
```diff
@@ -305,15 +331,97 @@ func (k Keeper) proofRequirementForClaim(ctx sdk.Context, claim *prooftypes.Clai
+
+		logger.Warn(fmt.Sprintf(
+			"unstaking supplier %s due to stake below the minimum",
+			supplierOperatorAddress,
```
let's log the owner as well

Olshansk commented on x/tokenomics/keeper/settle_pending_claims.go:
```diff
@@ -305,15 +331,97 @@ func (k Keeper) proofRequirementForClaim(ctx sdk.Context, claim *prooftypes.Clai
+			supplierOperatorAddress,
+		))
+
+		// TODO_CONSIDERATION: Should we just remove the supplier if the stake is
```
Yea, let's move this into a TODO_MAINNET

Olshansk commented on x/tokenomics/keeper/settle_pending_claims.go:
```diff
@@ -305,15 +331,97 @@ func (k Keeper) proofRequirementForClaim(ctx sdk.Context, claim *prooftypes.Clai
+		return err
+	}
+
+	// TODO_CONSIDERATION: Handle the case where the total slashing amount is
```
TODO_POST_MAINNET

bryanchriswhite commented on x/tokenomics/types/params.go:
```diff
@@ -52,6 +56,17 @@ func (params *Params) ValidateBasic() error {
 }
 
+// numComputeUnitsToCoin calculates the amount of uPOKT to mint based on the number of compute units.
+func (params *Params) NumComputeUnitsToCoin(numClaimComputeUnits uint64) (sdk.Coin, error) {
```
@red-0ne wdyt about refactoring it as a static function and adding it to somewhere like [https://github.com/pokt-network/poktroll/blob/main/x/shared/session.go](x/shared/session.go)?

red-0ne replied:
As discussed off-line, adding it to `x/shared` would create a cyclic dependency.

Adding it to `x/tokenomics/tokenomics.go` instead.

red-0ne commented on pkg/client/interface.go:
```diff
@@ -355,8 +356,20 @@ type TokenomicsRelayMiningDifficulty interface {
+// is necessary to prevent dependency cycles.
+type TokenomicsParams interface {
+	GetComputeUnitsToTokensMultiplier() uint64
+	NumComputeUnitsToCoin(numClaimComputeUnits uint64) (cosmostypes.Coin, error)
```
Adding it to `x/tokenomics/tokenomics.go`

Olshansk commented on x/tokenomics/keeper/settle_pending_claims.go:
```diff
@@ -305,15 +335,118 @@ func (k Keeper) proofRequirementForClaim(ctx sdk.Context, claim *prooftypes.Clai
+
+	supplierToSlash, supplierFound := k.supplierKeeper.GetSupplier(ctx, supplierOperatorAddress)
+	if !supplierFound {
+		return tokenomicstypes.ErrTokenomicsSupplierNotFound
```
Please wrap the error so it's clear it wasn't found for slashing purposes

Olshansk commented on x/tokenomics/keeper/settle_pending_claims.go:
```diff
@@ -305,15 +335,118 @@ func (k Keeper) proofRequirementForClaim(ctx sdk.Context, claim *prooftypes.Clai
+	if slashedSupplierInitialStakeCoin.IsGTE(totalSlashingCoin) {
+		remainingStakeCoin = slashedSupplierInitialStakeCoin.Sub(totalSlashingCoin)
+	} else {
+		logger.Warn(fmt.Sprintf(
```
```suggestion
		// TODO_MAINNET: Consider emitting an event for this situation.
		logger.Warn(fmt.Sprintf(
```

Olshansk commented on x/tokenomics/keeper/settle_pending_claims.go:
```diff
@@ -305,15 +335,118 @@ func (k Keeper) proofRequirementForClaim(ctx sdk.Context, claim *prooftypes.Clai
+
+	supplierToSlash.Stake = &remainingStakeCoin
+
+	logger.Warn(fmt.Sprintf(
```
This is not a warning. 

```suggestion
	logger.Info(fmt.Sprintf(
```

Olshansk commented on x/tokenomics/keeper/settle_pending_claims.go:
```diff
@@ -305,15 +335,118 @@ func (k Keeper) proofRequirementForClaim(ctx sdk.Context, claim *prooftypes.Clai
 	return requirementReason, nil
 }
 
+// slashSupplierStake slashes the stake of a supplier and mints the total slashing
+// amount to the tokenomics module account.
```
```suggestion
// slashSupplierStake slashes the stake of a supplier and transfers the total slashing applying
// from the supplier bank module to the tokenomics bank module account
```

Olshansk commented on x/tokenomics/tokenomics.go:
```diff
@@ -0,0 +1,20 @@
+	"github.com/pokt-network/poktroll/x/tokenomics/types"
+)
+
+// NumComputeUnitsToCoin calculates the amount of uPOKT to mint based on the number of compute units.
```
```suggestion
// NumComputeUnitsToCoin converts compute units to uPOKT based on global network parameters.
```

Olshansk commented on x/tokenomics/tokenomics.go:
```diff
@@ -0,0 +1,20 @@
+
+// NumComputeUnitsToCoin calculates the amount of uPOKT to mint based on the number of compute units.
+func NumComputeUnitsToCoin(params types.Params, numClaimComputeUnits uint64) (sdk.Coin, error) {
+	// CUPR is a LOCAL service specific parameter
```
```suggestion
	// CUTTM is a GLOBAL network wide parameter
```

red-0ne replied:
Good catch!

Olshansk commented on x/tokenomics/tokenomics.go:
```diff
@@ -0,0 +1,20 @@
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/pokt-network/poktroll/app/volatile"
+	"github.com/pokt-network/poktroll/x/tokenomics/types"
```
```suggestion
	tokenomicstypes "github.com/pokt-network/poktroll/x/tokenomics/types"
```

Olshansk commented on x/tokenomics/tokenomics.go:
```diff
@@ -0,0 +1,20 @@
+	// CUPR is a LOCAL service specific parameter
+	upoktAmount := math.NewInt(int64(numClaimComputeUnits * params.ComputeUnitsToTokensMultiplier))
+	if upoktAmount.IsNegative() {
+		return sdk.Coin{}, types.ErrTokenomicsRootHashInvalid.Wrap("sum * compute_units_to_tokens_multiplier is negative")
```
Please create a new error and update the message to reflect that fact this is an independent function
```suggestion
		return sdk.Coin{}, types.ErrTokenomicsCalculation.Wrap("compute_units * compute_units_to_tokens_multiplier is negative")
```

Olshansk commented on x/tokenomics/keeper/token_logic_modules.go:
```diff
@@ -20,6 +20,7 @@ import (
 	sessiontypes "github.com/pokt-network/poktroll/x/session/types"
 	sharedtypes "github.com/pokt-network/poktroll/x/shared/types"
 	suppliertypes "github.com/pokt-network/poktroll/x/supplier/types"
+	"github.com/pokt-network/poktroll/x/tokenomics"
```
Does your IDE not auto sort imports on save?

red-0ne replied:
Just checked, it actually sorted it there.

Olshansk commented on x/tokenomics/keeper/settle_pending_claims.go:
```diff
@@ -305,15 +335,118 @@ func (k Keeper) proofRequirementForClaim(ctx sdk.Context, claim *prooftypes.Clai
+	// Check if the supplier's stake is below the minimum and unstake it if necessary.
+	// TODO_BETA(@bryanchriswhite, #612): Use minimum stake governance parameter once available.
+	minSupplierStakeCoin := sdk.NewCoin(volatile.DenomuPOKT, math.NewInt(1))
+	// TODO_MAINNET(@red-0ne): Since SettlePendingClaims is not necessarily called
+	// at session end height so the unstaked supplier may not be immediately removed.
+	// Ensure that Gateways and Applications do not interact with a supplier that
+	// is below the minimum stake.
```
```suggestion
	// TODO_MAINNET(@red-0ne): SettlePendingClaims is called at the end of every block,
	// but not every block corresponds to the end of a session. This may lead to a situation
	// where a force unstaked supplier may still be able to interact with a Gateway or Application.
	// However, claims are only processed when sessions end.
	// INVESTIGATION: This requires an investigation if the race condition exists at all and fixed
	// only if it does.
```

Olshansk commented on x/tokenomics/keeper/settle_pending_claims.go:
```diff
@@ -305,15 +335,118 @@ func (k Keeper) proofRequirementForClaim(ctx sdk.Context, claim *prooftypes.Clai
+
+		// TODO_MAINNET: Should we just remove the supplier if the stake is
+		// below the minimum, at the risk of making the off-chain actors have an
+		// inconsistent session supplier list?
```
```suggestion
		// inconsistent session supplier list? See the comment above for more details.
```

red-0ne commented:
@Olshansk

I added an unexpected change due to a conflict (cyclic dependency) resulting from merging `main` into this branch.

The `ComputeUnitsToTokenMultiplier` is now moved to the `shared` module to resolve a conflict.

The conflict itself is due to `main` using `ProofRequirementForClaim` from the `proof` module.

While in this branch, `ProofRequirementForClaim` needs the `ComputeUnitsToTokenMultiplier` (from `tokenomics` module) to check against the `ProofRequirementThreshold` which results in a cyclic dependency.

Olshansk reviewed (approved):
Minor comment but LGTM otherwise!
